### PR TITLE
[ConstraintSystem] (mostly) Finish up solver type cache implementation

### PR DESCRIPTION
**Pull request by xedin**
Based on all of the great work @rudkx has done, solver should only apply _final_ types
to the expression tree when the solution is deduced. This patch makes it so and fixes
tests (mostly in IDE code completion) along the way.

### CODE REVIEW

**Comment by xedin** [comment]
@rudkx There is one thing left which is going to fix 2 crashers related to explicit casts - we still use `getCastTypeLoc().setType` in CSGen to set opened generic type.

**Comment by rudkx** [comment]
@xedin I didn't do any work on `TypeLoc` types yet. That seemed small enough to make it a separate task.

I haven't looked at this for a while, but I believe you also need to clear the type on the Expr when it is cached to ensure that we aren't accidentally reading it in places. That was the real hold up the last time I looked at this.

Can you hold off trying to merge this until I get done with the IUO changes. I suspect I've introduced some new `->getType()` and such that will need to be cleaned up with these changes in place, and I don't want to slow that work down at all.

**Comment by xedin** [comment]
@rudkx Sure, there is no rush in merging this, just wanted to get it out since tests are fixed. I has planning to take care of `getCastTypeLoc` before merging anyway.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

File: lib/Sema/CSDiag.cpp
Comment (rudkx) [inline lib/Sema/CSDiag.cpp]: Why is this directly calling `loc.getType()` and then only if getting `null` calling `cs.getType(loc)`?

File: test/Compatibility/tuple_arguments.swift
Comment (rudkx) [inline test/Compatibility/tuple_arguments.swift]: It seems like there is something wrong if this is resulting in this type of diagnostic change.

File: test/IDE/complete_constructor.swift
Comment (rudkx) [inline test/IDE/complete_constructor.swift]: So I am wondering how the code completion changes (which I haven't looked at closely) are related to the type cache. I would expect no behavioral changes from making the type checker store types in a different place.

File: lib/Sema/CSDiag.cpp
Comment (rudkx) [inline lib/Sema/CSDiag.cpp]: It's non-obvious why this is the right thing to do. This seems like an indication that something that should be reading from the type cache isn't.

File: lib/Sema/CSApply.cpp
Comment (rudkx) [inline lib/Sema/CSApply.cpp]: Under what circumstances do we need to do this? We should be type checking these bodies independently. They can refer to captured data, but that data should already have a type, right?

**Review by rudkx — CHANGES_REQUESTED** [review]
I'd like to have a better understanding of some of the changes and why we see behavioral changes here before we merge anything for this.

**Review by rudkx — APPROVED** [review]
We had a chance to sync up on this and everything looks okay for now. It definitely moves things forward re: fixing some crashers and no longer writing types all over the place. We still need to go back and clear expression types, etc. to finish this up.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please smoke test

**Pull request merged**
